data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/processor/peerforwarder/HashRing.java

```diff
@@ -0,0 +1,64 @@
+package com.amazon.dataprepper.plugins.processor.peerforwarder;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.NotThreadSafe;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.zip.CRC32;
+import java.util.zip.Checksum;
+
+@NotThreadSafe
+public class HashRing {
+    private final List<String> serverIps = new ArrayList<>();
+    private final int numVirtualNodes;
+    private final TreeMap<Long, String> virtualNodes = new TreeMap<>();
+
+    public HashRing(@Nonnull final List<String> serverIps, final int numVirtualNodes) {
+        this.numVirtualNodes = numVirtualNodes;
+        final List<String> sortedServerIps = new ArrayList<>(serverIps);
+        Collections.sort(sortedServerIps);
+        for (final String serverIp: sortedServerIps) {
+            addServerIp(serverIp);
+        }
+    }
+
+    public List<String> getServerIps() {
+        return serverIps;
+    }
+
+    private void addServerIp(@Nonnull final String serverIp) {
+        serverIps.add(serverIp);
+        final byte[] serverIpInBytes = serverIp.getBytes();
+        final Checksum crc32 = new CRC32();
+        final ByteBuffer intBuffer = ByteBuffer.allocate(4);
+        for (int i = 0; i < numVirtualNodes; i++) {
+            crc32.update(serverIpInBytes, 0, serverIpInBytes.length);
+            intBuffer.putInt(i);
+            crc32.update(intBuffer.array(), 0, intBuffer.array().length);
+            virtualNodes.putIfAbsent(crc32.getValue(), serverIp);
+            crc32.reset();
+            intBuffer.clear();
+        }
+    }
+
+    public String getServerIp(@Nonnull final byte[] traceId) {
```

Comment:
Any disadvantage to requiring a String and converting it to bytes for the caller here?

Reply:
It was easier to getTraceId as byte array from span. But since we already use hexString for traceId in the PeerForwarder, it is more readable to change it to String.